source/opt/vector_dce.cpp

```diff
@@ -0,0 +1,287 @@
+    WorkListItem new_item;
+    new_item.instruction = operand_inst;
+    new_item.components.Set(current_inst->GetSingleWordInOperand(1));
+    AddItemToWorkListIfNeeded(new_item, live_components, work_list);
+  }
+}
+
+void VectorDCE::MarkInsertUsesAsLive(
+    const VectorDCE::WorkListItem& current_item,
+    LiveComponentMap* live_components,
+    std::vector<VectorDCE::WorkListItem>* work_list) {
+  analysis::DefUseManager* def_use_mgr = context()->get_def_use_mgr();
+
+  uint32_t operand_id =
+      current_item.instruction->GetSingleWordInOperand(kInsertCompositeIdInIdx);
+  ir::Instruction* operand_inst = def_use_mgr->GetDef(operand_id);
+
+  WorkListItem new_item;
+  new_item.instruction = operand_inst;
+  new_item.components = current_item.components;
+  new_item.components.Clear(
+      current_item.instruction->GetSingleWordInOperand(2));
+
+  AddItemToWorkListIfNeeded(new_item, live_components, work_list);
+}
+
+void VectorDCE::MarkVectorShuffleUsesAsLive(
+    const WorkListItem& current_item,
+    VectorDCE::LiveComponentMap* live_components,
+    std::vector<WorkListItem>* work_list) {
+  analysis::DefUseManager* def_use_mgr = context()->get_def_use_mgr();
+
+  WorkListItem first_operand;
+  first_operand.instruction =
+      def_use_mgr->GetDef(current_item.instruction->GetSingleWordInOperand(0));
+  WorkListItem second_operand;
+  second_operand.instruction =
+      def_use_mgr->GetDef(current_item.instruction->GetSingleWordInOperand(1));
+
+  analysis::TypeManager* type_mgr = context()->get_type_mgr();
+  analysis::Vector* first_type =
+      type_mgr->GetType(first_operand.instruction->type_id())->AsVector();
+  uint32_t size_of_first_operand = first_type->element_count();
+
+  for (uint32_t in_op = 2; in_op < current_item.instruction->NumInOperands();
+       ++in_op) {
+    uint32_t index = current_item.instruction->GetSingleWordInOperand(in_op);
+    if (current_item.components.Get(in_op - 2)) {
+      if (index < size_of_first_operand) {
+        first_operand.components.Set(index);
+      } else {
+        second_operand.components.Set(index - size_of_first_operand);
+      }
+    }
+  }
+
+  AddItemToWorkListIfNeeded(first_operand, live_components, work_list);
+  AddItemToWorkListIfNeeded(second_operand, live_components, work_list);
+}
+
+void VectorDCE::MarkUsesAsLive(
+    ir::Instruction* current_inst, const utils::BitVector& live_elements,
+    LiveComponentMap* live_components,
+    std::vector<VectorDCE::WorkListItem>* work_list) {
+  analysis::DefUseManager* def_use_mgr = context()->get_def_use_mgr();
+
+  current_inst->ForEachInId([&work_list, &live_elements, this, live_components,
+                             def_use_mgr](uint32_t* operand_id) {
+    ir::Instruction* operand_inst = def_use_mgr->GetDef(*operand_id);
+
+    if (HasVectorResult(operand_inst)) {
+      WorkListItem new_item;
+      new_item.instruction = operand_inst;
+      new_item.components = live_elements;
```

Comment:
This seems wrong for OpCompositeConstruct. If I construct a float4 from 2 float2s where the only elements 2 and 3 are live, this should translate to two items: {0,0} and {1,1}, but it appears you generate {0,0,1,1} twice.

Reply:
This depends on what values are passed in for `live_elements`.  For the `OpCompositeConstruct` case (line 87), we pass in `all_components_live_`, which is 16 elements live for both inputs.  It is not as precise as it could be, which is why I left the TODO on line 85, but it is safe.